File: gcs/windows/src/BumblebeeGcs/Services/DroneApi.cs
// Copyright 2026 FutureLab
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Text;
using System.Text.Json;

namespace BumblebeeGcs.Services;

/// <summary>
/// HTTP-клиент для эндпоинтов дрона `/api/*` (бэкенд wifi_manager.py на :8765).
/// Покрывает LED, звуки/TTS (Entertainment) и wifi/auth (Settings).
/// </summary>
public sealed class DroneApi
{
    public const int Port = 8765;
    public string Host = "";

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(6) };

    public sealed class ApiException(string message) : Exception(message);

    public readonly record struct WifiStatus(string Mode, string Ssid, string Ip, int Signal);
    public readonly record struct ScanNet(string Ssid, int Signal, bool InUse, string Band, int Freq);

    private async Task<(byte[] Data, int Status)> RequestAsync(HttpMethod method, string path, object? json = null)
    {
        var host = Host;
        if (string.IsNullOrEmpty(host)) throw new ApiException("no drone");
        using var req = new HttpRequestMessage(method, $"http://{host}:{Port}{path}");
        if (json is not null)
            req.Content = new StringContent(JsonSerializer.Serialize(json), Encoding.UTF8, "application/json");
        HttpResponseMessage resp;
        try { resp = await Http.SendAsync(req); }
        catch (Exception ex) { throw new ApiException(ex.Message); }
        using (resp)
        {
            var data = await resp.Content.ReadAsByteArrayAsync();
            return (data, (int)resp.StatusCode);
        }
    }

    private async Task OkAsync(HttpMethod method, string path, object? json = null)
    {
        var (data, status) = await RequestAsync(method, path, json);
        if (status == 200) return;
        throw new ApiException(ErrorFrom(data) ?? $"HTTP {status}");
    }

    static string? ErrorFrom(byte[] data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind == JsonValueKind.Object &&
                doc.RootElement.TryGetProperty("error", out var err) && err.GetString() is string s)
                return s;
        }
        catch { }
        return null;
    }

    static Dictionary<string, JsonElement> ObjectFrom(byte[] data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind == JsonValueKind.Object)
                return doc.RootElement.EnumerateObject().ToDictionary(p => p.Name, p => p.Value.Clone());
        }
        catch { }
        return new Dictionary<string, JsonElement>();
    }

    // MARK: - LED

    public Task LedColorAsync(int r, int g, int b) => OkAsync(HttpMethod.Post, "/api/led/color", new { r, g, b });
    public Task LedAnimationAsync(string name) => OkAsync(HttpMethod.Post, "/api/led/animation", new { name });
    public Task LedResetAsync() => OkAsync(HttpMethod.Post, "/api/led/reset");

    // MARK: - Sounds / TTS

    /// <summary>Возвращает объект ответа сервера (может содержать "tune"/"phonemes").</summary>
    public async Task<Dictionary<string, JsonElement>> PlayTuneAsync(string tune, string format)
    {
        var (data, status) = await RequestAsync(HttpMethod.Post, "/api/sounds/play", new { tune, format });
        var obj = ObjectFrom(data);
        if (status != 200) throw new ApiException(Str(obj, "error") ?? $"HTTP {status}");
        return obj;
    }

    public async Task<Dictionary<string, JsonElement>> TtsAsync(string text, string lang)
    {
        var (data, status) = await RequestAsync(HttpMethod.Post, "/api/sounds/tts", new { text, lang });
        var obj = ObjectFrom(data);
        if (status != 200) throw new ApiException(Str(obj, "error") ?? $"HTTP {status}");
        return obj;
    }

    public static string? Str(Dictionary<string, JsonElement> obj, string key) =>
        obj.TryGetValue(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    // MARK: - WiFi / auth

    public async Task<WifiStatus> WifiStatusAsync()
    {
        var (data, _) = await RequestAsync(HttpMethod.Get, "/api/wifi/status");
        var o = ObjectFrom(data);
        return new WifiStatus(
            Str(o, "mode") ?? "—",
            Str(o, "ssid") ?? "—",
            Str(o, "ip") ?? "—",
            o.TryGetValue("signal", out var s) && s.ValueKind == JsonValueKind.Number ? (int)s.GetDouble() : 0);
    }

    public Task WifiSetModeAsync(string mode) => OkAsync(HttpMethod.Post, "/api/wifi/mode", new { mode });

    public async Task<bool> LoginAsync(string pin)
    {
        var (_, status) = await RequestAsync(HttpMethod.Post, "/api/auth/login", new { pin });
        return status == 200;
    }

    public async Task LogoutAsync()
    {
        try { await RequestAsync(HttpMethod.Post, "/api/auth/logout"); } catch { }
    }

    public async Task<List<ScanNet>> WifiScanAsync()
    {
        var (data, _) = await RequestAsync(HttpMethod.Get, "/api/wifi/scan");
        var list = new List<ScanNet>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return list;
            foreach (var el in doc.RootElement.EnumerateArray())
            {
                if (el.ValueKind != JsonValueKind.Object) continue;
                list.Add(new ScanNet(
                    el.TryGetProperty("ssid", out var ss) ? ss.GetString() ?? "" : "",
                    el.TryGetProperty("signal", out var sg) && sg.ValueKind == JsonValueKind.Number ? (int)sg.GetDouble() : 0,
                    el.TryGetProperty("in_use", out var iu) && iu.ValueKind == JsonValueKind.True,
                    el.TryGetProperty("band", out var bd) ? bd.GetString() ?? "" : "",
                    el.TryGetProperty("freq", out var fq) && fq.ValueKind == JsonValueKind.Number ? (int)fq.GetDouble() : 0));
            }
        }
        catch { }
        return list;
    }

    public async Task<List<string>> SavedNetworksAsync()
    {
        var (data, _) = await RequestAsync(HttpMethod.Get, "/api/wifi/networks");
        var list = new List<string>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return list;
            foreach (var el in doc.RootElement.EnumerateArray())
                if (el.ValueKind == JsonValueKind.Object &&
                    el.TryGetProperty("name", out var n) && n.GetString() is string name)
                    list.Add(name);
        }
        catch { }
        return list;
    }

    public Task RemoveNetworkAsync(string ssid) => OkAsync(HttpMethod.Delete, "/api/wifi/network", new { ssid });
    public Task SaveNetworkAsync(string ssid, string password) => OkAsync(HttpMethod.Post, "/api/wifi/save", new { ssid, password });
    public Task ApplyNowAsync(string ssid) => OkAsync(HttpMethod.Post, "/api/wifi/apply-now", new { ssid });
    public Task ApplyBootAsync(string ssid) => OkAsync(HttpMethod.Post, "/api/wifi/apply-boot", new { ssid });
}

File: gcs/macos/Sources/GCSApp/DroneAPI.swift
import Foundation

/// HTTP client for the drone's `/api/*` endpoints, served directly by the
/// on-drone backend (`wifi_manager.py`) on :8765. The web dashboard (nginx :80)
/// has been removed; the app talks to 8765 straight, no reverse proxy.
/// Covers LED, sounds/TTS (Entertainment) and wifi/auth (Settings).
final class DroneAPI {
    /// Backend port that `wifi_manager.py` binds (previously behind nginx :80).
    static let port = 8765
    var host = ""
    private let session = URLSession(configuration: .ephemeral)

    struct APIError: LocalizedError { let message: String; var errorDescription: String? { message } }

    private func url(_ path: String) -> URL? { URL(string: "http://\(host):\(Self.port)\(path)") }

    // Generic request returning raw Data + status.
    private func request(_ method: String, _ path: String,
                         json: [String: Any]? = nil) async throws -> (Data, Int) {
        guard !host.isEmpty, let url = url(path) else { throw APIError(message: "no drone") }
        var req = URLRequest(url: url, timeoutInterval: 6.0)
        req.httpMethod = method
        if let json = json {
            req.httpBody = try JSONSerialization.data(withJSONObject: json)
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, resp) = try await session.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        return (data, status)
    }

    @discardableResult
    private func ok(_ method: String, _ path: String, json: [String: Any]? = nil) async throws -> Bool {
        let (data, status) = try await request(method, path, json: json)
        if status == 200 { return true }
        if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let err = obj["error"] as? String {
            throw APIError(message: err)
        }
        throw APIError(message: "HTTP \(status)")
    }

    private func jsonObject(_ method: String, _ path: String,
                            json: [String: Any]? = nil) async throws -> Any {
        let (data, _) = try await request(method, path, json: json)
        return try JSONSerialization.jsonObject(with: data)
    }

    // MARK: - LED

    func ledColor(r: Int, g: Int, b: Int) async throws {
        try await ok("POST", "/api/led/color", json: ["r": r, "g": g, "b": b])
    }
    func ledAnimation(_ name: String) async throws {
        try await ok("POST", "/api/led/animation", json: ["name": name])
    }
    func ledReset() async throws {
        try await ok("POST", "/api/led/reset")
    }

    // MARK: - Sounds / TTS

    /// Returns the server response object (may contain "tune"/"phonemes").
    @discardableResult
    func playTune(_ tune: String, format: String) async throws -> [String: Any] {
        let (data, status) = try await request("POST", "/api/sounds/play",
                                               json: ["tune": tune, "format": format])
        let obj = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
        if status != 200 { throw APIError(message: (obj["error"] as? String) ?? "HTTP \(status)") }
        return obj
    }

    @discardableResult
    func tts(text: String, lang: String) async throws -> [String: Any] {
        let (data, status) = try await request("POST", "/api/sounds/tts",
                                               json: ["text": text, "lang": lang])
        let obj = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
        if status != 200 { throw APIError(message: (obj["error"] as? String) ?? "HTTP \(status)") }
        return obj
    }

    // MARK: - WiFi / auth

    struct WifiStatus { var mode: String; var ssid: String; var ip: String; var signal: Int }
    struct ScanNet: Identifiable { let id = UUID(); var ssid: String; var signal: Int; var inUse: Bool; var band: String; var freq: Int }

    func wifiStatus() async throws -> WifiStatus {
        let obj = (try await jsonObject("GET", "/api/wifi/status")) as? [String: Any] ?? [:]
        return WifiStatus(mode: obj["mode"] as? String ?? "—",
                          ssid: obj["ssid"] as? String ?? "—",
                          ip: obj["ip"] as? String ?? "—",
                          signal: obj["signal"] as? Int ?? 0)
    }
    func wifiSetMode(_ mode: String) async throws { try await ok("POST", "/api/wifi/mode", json: ["mode": mode]) }
    func login(pin: String) async throws -> Bool {
        let (_, status) = try await request("POST", "/api/auth/login", json: ["pin": pin])
        return status == 200
    }
    func logout() async throws { _ = try? await request("POST", "/api/auth/logout") }

    func wifiScan() async throws -> [ScanNet] {
        let arr = (try await jsonObject("GET", "/api/wifi/scan")) as? [[String: Any]] ?? []
        return arr.map {
            ScanNet(ssid: $0["ssid"] as? String ?? "",
                    signal: $0["signal"] as? Int ?? 0,
                    inUse: $0["in_use"] as? Bool ?? false,
                    band: $0["band"] as? String ?? "",
                    freq: $0["freq"] as? Int ?? 0)
        }
    }
    func savedNetworks() async throws -> [String] {
        let arr = (try await jsonObject("GET", "/api/wifi/networks")) as? [[String: Any]] ?? []
        return arr.compactMap { $0["name"] as? String }
    }
    func removeNetwork(_ ssid: String) async throws { try await ok("DELETE", "/api/wifi/network", json: ["ssid": ssid]) }
    func saveNetwork(ssid: String, password: String) async throws { try await ok("POST", "/api/wifi/save", json: ["ssid": ssid, "password": password]) }
    func applyNow(ssid: String) async throws { try await ok("POST", "/api/wifi/apply-now", json: ["ssid": ssid]) }
    func applyBoot(ssid: String) async throws { try await ok("POST", "/api/wifi/apply-boot", json: ["ssid": ssid]) }
}
